keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+
+
+By design, DRA uses parameters for resources that are completely
+opaque to core Kubernetes. They get interpreted by a DRA driver's controller
+(for allocating claims) and a DRA driver's kubelet plugin (for configuring
+resources a node). During scheduling of a pod, the kube-scheduler and any DRA
+driver controller(s) handling claims for the pod communicate back-and-forth through the
+apiserver by updating a `PodSchedulingContext` object, ultimately leading to the
+allocation of all pending claims and the pod being scheduled onto a node.
+
+This approach poses a problem for the [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler) (CA) or for any higher
+level controller that needs to make decisions for a group of pods (e.g. a job
+scheduler). It cannot simulate the effect of allocating or deallocating
+claims over time. Only the third-party DRA
+drivers have the information available to do this.
+
+"Semantic parameters" is an extension to DRA that addresses this problem by
+making claim parameters less opaque. Instead of managing the semantics of all
+claim parameters themselves, drivers now manage resources and describe them
+using a specific "semantic model" pre-defined by Kubernetes. This allows
+components aware of this "semantic model" to make decisions about these
+resources without outsourcing them to some third-party controller. For example,
+the scheduler is now able to allocate claims rapidly, without back-and-forth
+communication with DRA drivers.
+
+At a high-level, this extension takes the following form:
+
+* DRA drivers publish their available resources in the form of a
+  `NodeResourceSlice` object on a node-by-node basis according to one of the
+  builtin "semantic models" known to Kubernetes. This object is stored in the
+  API server and available to the scheduler (or Cluster Autoscaler) to query
+  when a resource request comes in later on.
+
+* When a user wants to consume a resource, they create a `ResourceClaim`,
+  which, in turn, references a claim parameters object. This object defines how
+  many resources are needed and which capabilities they must have. Typically, it
+  is defined using a vendor-specific type which might also support configuration
+  parameters (i.e. parameters that are *not* needed for allocation but *are*
+  needed for configuration).
+
+* With such a claim in place, DRA drivers "resolve" the contents of any
+  vendor-specific claim parameters into a canonical form (i.e. a generic
+  `ResourceClaimParameters` object in the `resource.k8s.io` API group) which
+  the scheduler (or Cluster Autoscaler) can evaluate against the
+  `NodeResourceSlice` of any candidate nodes without knowing exactly what is
+  being requested. They then use this information to help decide which node to
+  schedule a pod on (as well as allocate resources from its `NodeResourceSlice`
+  in the process).
+
+* Once a node is chosen and a workload is sent to it, DRA drivers are
```

Comment:
I might say something a little bit more here, to note that the allocation decision is stored and used by the scheduler for future allocations, even before actuation of the decision is complete. Something like:

"Once a node is chosen and the allocation decisions made, the scheduler will store the result in the API server as well as update its in-memory model of available resources. DRA drivers are responsible for using this allocation result to inject any allocated resources..."

Reply:
Changed.